From 73249ef8bfbff206df0b07b03b404e921001c9b4 Mon Sep 17 00:00:00 2001
From: rpkak <rpkak@users.noreply.github.com>
Date: Tue, 23 Sep 2025 12:58:08 +0200
Subject: [PATCH 114/133] optimize std.mem.swap
X-Developer-Signature: v=1; a=openpgp-sha256; l=1696; i=xtex@astrafall.org;
 h=from:subject; bh=Sji1cqP9MbMooqM28nQ2HMVRvbwB0XTRjEpHgzPs6WA=;
 b=owGbwMvMwCW2U4Ij7wZL9ETG02pJDJn78niZGE0FrXmfxAktWekfvkGl6NX6okurQiOSTngJG
 epvlJfqmMjCIMbFYCmmyFJk2ODNqpPOL7qsXBZmDisTyBBpkQYGBgYGFga+3MS8UiMdIz1TbUM9
 Q0MdIx0jBi5OAZjqp2cYGc6XFt1uvvim5uMJ9h+yW9jcJqX+CVfYbH5M9J2PW1zj1VBGhpUTeh/
 UvA4SCFWz2HbYTIRht5oj+74pH0LTb5Rmz9mVxAQA
X-Developer-Key: i=xtex@astrafall.org; a=openpgp;
 fpr=7231804B052C670F15A6771DB918086ED8045B91

(cherry picked from commit 4fb08986cba65623c6dbabcbb83f0aa570e06a65)
---
 lib/std/mem.zig | 32 ++++++++++++++++++++++++++++----
 1 file changed, 28 insertions(+), 4 deletions(-)

diff --git a/lib/std/mem.zig b/lib/std/mem.zig
index 1a67888c63bf..6f06d45a8ed5 100644
--- a/lib/std/mem.zig
+++ b/lib/std/mem.zig
@@ -3643,10 +3643,34 @@ test indexOfMinMax {
 }
 
 /// Exchanges contents of two memory locations.
-pub fn swap(comptime T: type, a: *T, b: *T) void {
-    const tmp = a.*;
-    a.* = b.*;
-    b.* = tmp;
+pub fn swap(comptime T: type, noalias a: *T, noalias b: *T) void {
+    if (@inComptime()) {
+        // In comptime, accessing bytes of values with no defined layout is a compile error.
+        const tmp = a.*;
+        a.* = b.*;
+        b.* = tmp;
+    } else {
+        // Swapping in streaming nature from start to end instead of swapping
+        // everything in one step allows easier optimizations and less stack usage.
+        const a_bytes: []align(@alignOf(T)) u8 = @ptrCast(a);
+        const b_bytes: []align(@alignOf(T)) u8 = @ptrCast(b);
+        for (a_bytes, b_bytes) |*ab, *bb| {
+            const tmp = ab.*;
+            ab.* = bb.*;
+            bb.* = tmp;
+        }
+    }
+}
+
+test "swap works at comptime with types with no defined layout" {
+    comptime {
+        const T = struct { val: u64 };
+        var a: T = .{ .val = 0 };
+        var b: T = .{ .val = 1 };
+        swap(T, &a, &b);
+        try testing.expectEqual(T{ .val = 1 }, a);
+        try testing.expectEqual(T{ .val = 0 }, b);
+    }
 }
 
 inline fn reverseVector(comptime N: usize, comptime T: type, a: []T) [N]T {
-- 
2.52.0

